db/range_del_aggregator.cc

```diff
@@ -9,233 +9,146 @@
-  assert(IsValueType(parsed.type));
-  assert(rep_ != nullptr);
-  auto& positional_tombstone_map = GetPositionalTombstoneMap(parsed.sequence);
-  const auto& tombstone_map = positional_tombstone_map.raw_map;
-  if (tombstone_map.empty()) {
-    return false;
-  }
-  auto& tombstone_map_iter = positional_tombstone_map.iter;
-  if (tombstone_map_iter == tombstone_map.end() &&
-      (mode == RangeDelPositioningMode::kForwardTraversal ||
-       mode == RangeDelPositioningMode::kBackwardTraversal)) {
-    // invalid (e.g., if AddTombstones() changed the deletions), so need to
-    // reseek
-    mode = RangeDelPositioningMode::kBinarySearch;
-  }
-  switch (mode) {
-    case RangeDelPositioningMode::kFullScan:
-      assert(!collapse_deletions_);
-      // The maintained state (PositionalTombstoneMap::iter) isn't useful when
-      // we linear scan from the beginning each time, but we maintain it anyways
-      // for consistency.
-      tombstone_map_iter = tombstone_map.begin();
-      while (tombstone_map_iter != tombstone_map.end()) {
-        const auto& tombstone = tombstone_map_iter->second;
-        if (icmp_.user_comparator()->Compare(parsed.user_key,
-                                             tombstone.start_key_) < 0) {
-          break;
-        }
-        if (parsed.sequence < tombstone.seq_ &&
-            icmp_.user_comparator()->Compare(parsed.user_key,
-                                             tombstone.end_key_) < 0) {
-          return true;
-        }
-        ++tombstone_map_iter;
-      }
-      return false;
-    case RangeDelPositioningMode::kForwardTraversal:
-      assert(collapse_deletions_ && tombstone_map_iter != tombstone_map.end());
-      if (tombstone_map_iter == tombstone_map.begin() &&
-          icmp_.user_comparator()->Compare(parsed.user_key,
-                                           tombstone_map_iter->first) < 0) {
-        // before start of deletion intervals
-        return false;
-      }
-      while (std::next(tombstone_map_iter) != tombstone_map.end() &&
-             icmp_.user_comparator()->Compare(
-                 std::next(tombstone_map_iter)->first, parsed.user_key) <= 0) {
-        ++tombstone_map_iter;
-      }
-      break;
-    case RangeDelPositioningMode::kBackwardTraversal:
-      assert(collapse_deletions_ && tombstone_map_iter != tombstone_map.end());
-      while (tombstone_map_iter != tombstone_map.begin() &&
-             icmp_.user_comparator()->Compare(parsed.user_key,
-                                              tombstone_map_iter->first) < 0) {
-        --tombstone_map_iter;
-      }
-      if (tombstone_map_iter == tombstone_map.begin() &&
-          icmp_.user_comparator()->Compare(parsed.user_key,
-                                           tombstone_map_iter->first) < 0) {
-        // before start of deletion intervals
-        return false;
+// An UncollapsedRangeDelMap is quick to create but slow to answer ShouldDelete
+// queries.
```

Comment:
Is it quicker to create than a `CollapsedRangeDelMap`? I don't think it is. My understanding is that an `UncollapsedRangeDelMap` is needed during sstable ingestion, but that understanding is fuzzy. Do you have more details to add here?

Reply:
Performance is the whole point of having an uncollapsed mode. If it's not actually faster we should just get rid of the uncollapsed mode. It's faster to build if you've got lots of overlapping ranges, but perhaps that doesn't matter when there are no lurking n^2 bugs in the overlapped mode.

Adding an SST calls IsRangeOverlapped, which was only implemented for the uncollapsed mode because it just wasn't needed for the collapsed mode. There's no reason it couldn't be implemented for collapsed mode.